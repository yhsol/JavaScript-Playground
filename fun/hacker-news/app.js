const container = document.getElementById('root');
const ajax = new XMLHttpRequest();
const content = document.createElement('div');
const NEWS_URL = 'https://api.hnpwa.com/v0/news/1.json';
const CONTENT_URL = 'https://api.hnpwa.com/v0/item/@id.json';

const store = {
    currentPage: 1,
}

function getData(url) {
    ajax.open('GET', url, false);
    ajax.send();
    
    return JSON.parse(ajax.response);
}

function newsFeed() {
    const newsFeed = getData(NEWS_URL);
    const newsList = [];

    let template = `
        <div>
            <h1>Hacker News</h1>
            <ul>
                {{__news_feed__}}
            </ul>
            <div>
                <a href="#/page/{{__prev_page__}}">prev</a>
                <a href="#/page/{{__next_page__}}">next</a>
            </div>
        </div>
    `;

    newsList.push('<ul>')
    for (let i = (store.currentPage - 1) * 10; i < store.currentPage * 10; i++) {
        newsList.push(`
        <li>
            <a href="#/show/${newsFeed[i].id}">
                ${newsFeed[i].title} (${newsFeed[i].comments_count})
            </a>
        </li>
        `);
    }
    newsList.push('</ul>')
    newsList.push(`
    <div>
        <a href="#/page/${store.currentPage > 1 ? store.currentPage - 1 : 1}">prev</a>
        <a href="#/page/${store.currentPage + 1}">next</a>
    </div>
    `);
    container.innerHTML = newsList.join('');
}

function newsDetail() {
    const newsContent = getData(CONTENT_URL.replace('@id', location.hash.substr(7)));
    
    container.innerHTML = `
    <h1>${newsContent.title}</h1>

    <div>
        <a href="#/page/${store.currentPage}">to list</a>
    </div>    
    `;
}

function router() {
    const routerPath = location.hash;
    
    if (routerPath === '') {
        newsFeed();
    } else if (routerPath.indexOf('#/page') >= 0) {
        store.currentPage = Number(routerPath.substr(7));
        newsFeed();
    } else {
        newsDetail();
    }

}

window.addEventListener('hashchange', router)

router();