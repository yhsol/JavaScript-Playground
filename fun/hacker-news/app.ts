type Store = {
    currentPage: number,
    feeds: NewsFeed[],
}

type News = {
    id: number,
    title: string,
    time_ago: string,
    url: string;
    user: string;
    content: string;
}

type NewsFeed = News & {
    readonly points: number,
    comments_count: number,
    read?: boolean,
}

type NewsDetail = News & {
    comments: NewsComment[];
}

type NewsComment = News & {
    level: number,
    comments: NewsComment[],
}

const container: HTMLElement | null = document.getElementById('root');
const ajax: XMLHttpRequest = new XMLHttpRequest();
const content: HTMLDivElement = document.createElement('div');
const NEWS_URL = 'https://api.hnpwa.com/v0/news/1.json';
const CONTENT_URL = 'https://api.hnpwa.com/v0/item/@id.json';

const store: Store = {
    currentPage: 1,
    feeds: [],
}

class Api {
    getRequest<T>(url: string): T {
        const ajax = new XMLHttpRequest();
        ajax.open('GET', url, false);
        ajax.send();
        
        return JSON.parse(ajax.response);
    }
}

// targetClass 로 제공된 class 에 baseClasses 로 제공된 class 들의 기능을 합성한다.
function applyApiMixins(targetClass: any, baseClasses: any[]) {
    baseClasses.forEach(baseClass => {
        Object.getOwnPropertyNames(baseClass.prototype).forEach(name => {
            const descriptor = Object.getOwnPropertyDescriptor(baseClass.prototype, name);

            if (descriptor) {
                Object.defineProperty(targetClass.prototype, name, descriptor);
            }
        })
    })
}

interface NewsFeedApi extends Api {};
interface NewsDetailApi extends Api {};

class NewsFeedApi {
    getData(): NewsFeed[] {
        return this.getRequest<NewsFeed[]>(NEWS_URL);
    }
}

class NewsDetailApi {
    getData(id: string): NewsDetail {
        return this.getRequest<NewsDetail>(CONTENT_URL.replace('@id', id));
    }
}

applyApiMixins(NewsFeedApi, [Api]);
applyApiMixins(NewsDetailApi, [Api]);

// function getData<AjaxResponse>(url: string): AjaxResponse {
//     ajax.open('GET', url, false);
//     ajax.send();
    
//     return JSON.parse(ajax.response);
// }

class NewsFeedView {
    constructor() {
        
    }
}

function makeFeeds(feeds: NewsFeed[]) {
    for (let i = 0; i < feeds.length; i++) {
        feeds[i].read = false;
    }

    return feeds;
}

function updateView(html: string) {
    if (container) {
        container.innerHTML = html;
    } else {
        console.log('container is null');
    }
}

function newsFeed() {
    const api = new NewsFeedApi();
    let newsFeeds: NewsFeed[] = store.feeds; // getData(NEWS_URL);
    const newsList = [];

    if (newsFeeds.length === 0) {
        newsFeeds = store.feeds = makeFeeds(api.getData());
    }

    // 변경할 부분을 마킹해 두고 template.replace 를 통해 변경한다.
    let template = `
        <div class="bg-gray-600 min-h-screen">
            <div class="bg-white text-xl">
                <div class="mx-auto px-4">
                    <div class="flex justify-between items-center py-6">
                        <div class="flex justify-start">
                            <h1 class="font-extrabold">Hacker News</h1>
                        </div>
                        <div class="items-center justify-end">
                            <a href="#/page/{{__prev_page__}}" class="text-gray-500">
                                Previous
                            </a>
                            <a href="#/page/{{__next_page__}}" class="text-gray-500">
                                Next
                            </a>
                        </div>
                    </div>
                </div>
            </div>
            <div class="p-4 text-2xl text-gray-700">
                {{__news_feed__}}
            </div>
        </div>
    `;

    for (let i = (store.currentPage - 1) * 10; i < store.currentPage * 10; i++) {
        newsList.push(`
        <div class="p-6 ${newsFeeds[i].read ? 'bg-gray-200' : 'bg-white'} mt-6 rounded-lg shadow-md transition-colors duration-500 hover:bg-green-100">
            <div class="flex">
                <div class="flex-auto">
                    <a href="#/show/${newsFeeds[i].id}">${newsFeeds[i].title}</a>
                </div>
                <div class="text-center text-sm">
                    <div class="w-10 text-white bg-green-300 rounded-lg px-0 py-2">${newsFeeds[i].comments_count}</div>
                </div>
            </div>
            <div class="flex mt-3">
                <div class="grid grid-cols-3 text-sm text-gray-500">
                    <div><i class="fas fa-user mr-1"></i>${newsFeeds[i].user}</div>
                    <div><i class="fas fa-heart mr-1"></i>${newsFeeds[i].points}</div>
                    <div><i class="fas fa-clock mr-1"></i>${newsFeeds[i].time_ago}</div>
                </div>
            </div>
        </div>
        `);
    }

    template = template.replace('{{__news_feed__}}', newsList.join(''));
    template = template.replace('{{__prev_page__}}', String(store.currentPage > 1 ? store.currentPage - 1 : 1));
    template = template.replace('{{__next_page__}}', String(store.currentPage +1));

    updateView(template);
}

function newsDetail() {
    const id = location.hash.substr(7)
    const api = new NewsDetailApi();
    const newsContent = api.getData(id);

    let template = `
        <div class="bg-gray-600 min-h-screen pb-8">
            <div class="bg-white text-xl">
                <div class="mx-auto px-4">
                    <div class="flex justify-between items-center py-6">
                        <div class="flex justify-start">
                            <h1 class="font-extrabold">Hacker News</h1>
                        </div>
                        <div class="items-center justify-end">
                            <a href="#/page/${store.currentPage}" class="text-gray-500">
                                <i class="fa fa-times"></i>
                            </a>
                        </div>
                    </div>
                </div>
            </div>

            <div class="h-full border rounded-xl bg-white m-6 p-4">
                <h2>${newsContent.title}</h2>
                <div class="text-gray-400 h-20">
                    ${newsContent.content}
                </div>

                {{__comments__}}

            </div>
        </div>
    `;

    for (let i = 0; i < store.feeds.length; i++) {
        if (store.feeds[i].id === Number(id)) {
            store.feeds[i].read = true;
            break;
        }
    }

    updateView(template.replace('{{__comments__}}', makeComment(newsContent.comments)));
}

function makeComment(comments: NewsComment[]) {
    const commentString: string[] = [];
    for (let i = 0; i < comments.length; i++) {
        const comment = comments[i];
        commentString.push(`
            <div style="padding-left: ${40 * comment.level}px;" class="mt-4">
                <div class="text-gray-400">
                    <i class="fa fa-sort-up mr-2"></i>
                    <strong>${comment.user}</strong> ${comment.time_ago}
                </div>
                <p class="text-gray-700">${comment.content}</p>
            </div>
        `);

        // comments 를 다시 호출하여 대댓글을 출력한다.
        if (comment.comments.length > 0) {
            commentString.push(makeComment(comment.comments));
        }
    }

    return commentString.join('');
}

function router() {
    const routerPath = location.hash;
    
    if (routerPath === '') {
        newsFeed();
    } else if (routerPath.indexOf('#/page') >= 0) {
        store.currentPage = Number(routerPath.substr(7));
        newsFeed();
    } else {
        newsDetail();
    }

}

window.addEventListener('hashchange', router)

router();