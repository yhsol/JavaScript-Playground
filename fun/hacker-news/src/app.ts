import Router from "./core/router";
import NewsDetailView from "./page/news-detail-view";
import NewsFeedView from "./page/news-feed-view";
import Store from "./store";

const store = new Store();
const router = new Router();
const newsFeedView = new NewsFeedView('root', store);
const newsDetailView = new NewsDetailView('root', store);

router.setDefaultView(newsFeedView);
router.addRouterPath('/page/', newsFeedView);
router.addRouterPath('/show/', newsDetailView);

router.route();