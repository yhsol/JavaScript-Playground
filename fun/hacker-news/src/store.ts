import { NewsFeed, NewsStore } from "./types"

export class Store implements NewsStore {
  private feeds: NewsFeed[];
  private _currentPage: number;

  constructor() {
    this.feeds = [];
    this._currentPage = 1;
  }

  get currentPage() {
    return this._currentPage;
  }

  set currentPage(page: number) {
    this._currentPage = page;
  }

  get nextPage() {
    return this._currentPage + 1;
  }

  get prevPage() {
    return this._currentPage > 1 ? this._currentPage - 1 : 1;
  }
  
  get numberOfFeed() {
    return this.feeds.length;
  }

  get hasFeeds() {
    return this.feeds.length > 0;
  }

  getAllFeeds() {
    return this.feeds;
  }

  getFeed(position: number) {
    return this.feeds[position];
  }

  setFeeds(feeds: NewsFeed[]) {
    this.feeds = feeds.map(feed => ({
      ...feed, 
      read: false,
    }));
  }

  makeRead(id: number) {
    const feed = this.feeds.find(feed => feed.id === id);

    if (feed) {
      feed.read = true;
    }
  }
}