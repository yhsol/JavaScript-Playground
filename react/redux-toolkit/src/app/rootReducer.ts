import { combineReducers } from "redux";
import profileReducer from "../features/profile/profile.slice";
import todoReducer from "../features/todo/todo.slice";

const moduleStores = {
  profile: profileReducer,
  todo: todoReducer,
};

const rootReducer = combineReducers({
  ...moduleStores,
});

export type RootState = ReturnType<typeof rootReducer>;

export default rootReducer;
