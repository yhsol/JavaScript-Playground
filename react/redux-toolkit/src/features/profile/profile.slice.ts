import { createSlice } from "@reduxjs/toolkit";
import {
  createProfileReducer,
  deleteProfileReducer,
  resetReducer,
  setActiveProfileReducer,
  updateProfileReducer,
} from "./reducers/profile.reducer";

const profileTypes = ["guest", "user", "admin"] as const;
export type ProfileType = typeof profileTypes[number];

export interface Profile {
  profileId: number;
  name?: string;
  profileType: ProfileType;
  showTodoList: boolean;
}

export interface ProfileState {
  profileList: Profile[];
  maxProfileId: number;
  profileTypes: ReadonlyArray<ProfileType>;
  activeProfileId?: number;
}

export const initialState: ProfileState = {
  profileList: [
    { profileId: 1, name: "Ben", profileType: "admin", showTodoList: true },
    { profileId: 2, name: "Sue", profileType: "user", showTodoList: false },
  ],
  maxProfileId: 2,
  profileTypes: profileTypes as ReadonlyArray<ProfileType>,
};

export const profileSlice = createSlice({
  name: "profile",
  initialState,
  reducers: {
    createProfile: createProfileReducer,
    deleteProfile: deleteProfileReducer,
    updateProfile: updateProfileReducer,
    setActiveProfile: setActiveProfileReducer,
    reset: resetReducer,
  },
});

export const {
  createProfile,
  deleteProfile,
  updateProfile,
  setActiveProfile,
  reset,
} = profileSlice.actions;

const profileReducer = profileSlice.reducer;
export default profileReducer;
