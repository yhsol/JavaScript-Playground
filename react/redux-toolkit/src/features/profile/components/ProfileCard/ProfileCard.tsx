import React from "react";
import { useSelector } from "react-redux";
import CardFooterWrapper from "../../../../app/component/CardFooterWrapper/CardFooterWrapper";
import { profileListSelector } from "../../selectors/profile.selector";
import ProfileItem from "../ProfileItem/ProfileItem";
import { getCreateOnClick } from "../ProfileItem/profileItem.funcs";

const ProfileCard = () => {
  const profileList = useSelector(profileListSelector);
  return (
    <div className="card" role="main" aria-label="Profile Card">
      <div className="card-header">
        <h5 className="card-title">Profiles</h5>
      </div>
      <div className="card-body">
        <ul className="list-group">
          {profileList.map((profile) => (
            <ProfileItem
              key={profile.profileId}
              profile={profile}
            ></ProfileItem>
          ))}
        </ul>
      </div>
      <div className="card-footer">
        <CardFooterWrapper>
          <button className="btn btn-primary" onClick={getCreateOnClick()}>
            Create New Profile
          </button>
        </CardFooterWrapper>
      </div>
    </div>
  );
};

export default ProfileCard;
