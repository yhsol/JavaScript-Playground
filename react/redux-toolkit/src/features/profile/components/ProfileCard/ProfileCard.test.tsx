import {
  fireEvent,
  render,
  RenderResult,
  screen,
  within,
} from "@testing-library/react";
import { Provider } from "react-redux";

import { store } from "../../../../app/store";
import { reset } from "../../profile.slice";
import ProfileCard from "./ProfileCard";

const renderCard = (): RenderResult =>
  render(
    <Provider store={store}>
      <ProfileCard />
    </Provider>
  );

test("renders the Profile Card", () => {
  const { asFragment } = renderCard();
  expect(asFragment()).toMatchSnapshot();
  expect(screen.getByLabelText(/profile card/i)).toBeInTheDocument();
  const profileList = screen.getByRole("list");
  expect(profileList).toBeInTheDocument();
  const profileListItems = within(profileList).getAllByRole("listitem");
  expect(profileListItems).toHaveLength(2);
});

test("test delete button", () => {
  const { rerender, asFragment } = renderCard();
  const profileList = screen.getByRole("list");
  let profileListItems = within(profileList).getAllByRole("listitem");
  const deleteButton = within(profileListItems[0]).getByText(/delete/i);
  fireEvent.click(deleteButton);
  rerender(
    <Provider store={store}>
      <ProfileCard />
    </Provider>
  );
  profileListItems = within(profileList).getAllByRole("listitem");
  expect(profileListItems).toHaveLength(1);
  expect(asFragment()).toMatchSnapshot();
});

test("test create button", () => {
  store.dispatch(reset());
  const { rerender, asFragment } = renderCard();
  const createButton = screen.getByText(/create/i);
  const profileList = screen.getByRole("list");
  let profileListItems = within(profileList).getAllByRole("listitem");
  expect(profileListItems).toHaveLength(2);
  fireEvent.click(createButton);
  rerender(
    <Provider store={store}>
      <ProfileCard />
    </Provider>
  );
  profileListItems = within(profileList).getAllByRole("listitem");
  expect(profileListItems).toHaveLength(3);
  expect(asFragment()).toMatchSnapshot();
  const lastItem = profileListItems[2];
  expect(within(lastItem).getByLabelText("Name")).toHaveValue("");
  expect(within(lastItem).getByLabelText("Type")).toHaveValue("guest");
});
