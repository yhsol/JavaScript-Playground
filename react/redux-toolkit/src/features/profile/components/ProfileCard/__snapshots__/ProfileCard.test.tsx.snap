// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Profile Card 1`] = `
<DocumentFragment>
  <div
    aria-label="Profile Card"
    class="card"
    role="main"
  >
    <div
      class="card-header"
    >
      <h5
        class="card-title"
      >
        Profiles
      </h5>
    </div>
    <div
      class="card-body"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item profileItem"
        >
          <div>
            <label
              for="name-1-input"
            >
              Name
            </label>
            <input
              class="form-control"
              id="name-1-input"
              type="text"
              value="Ben"
            />
          </div>
          <div>
            <label
              for="type-1-select"
            >
              Type
            </label>
            <select
              class="form-control"
              id="type-1-select"
            >
              <option>
                guest
              </option>
              <option>
                user
              </option>
              <option>
                admin
              </option>
            </select>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item profileItem"
        >
          <div>
            <label
              for="name-2-input"
            >
              Name
            </label>
            <input
              class="form-control"
              id="name-2-input"
              type="text"
              value="Sue"
            />
          </div>
          <div>
            <label
              for="type-2-select"
            >
              Type
            </label>
            <select
              class="form-control"
              id="type-2-select"
            >
              <option>
                guest
              </option>
              <option>
                user
              </option>
              <option>
                admin
              </option>
            </select>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    >
      <div
        class="wrapper"
      >
        <button
          class="btn btn-primary"
        >
          Create New Profile
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test create button 1`] = `
<DocumentFragment>
  <div
    aria-label="Profile Card"
    class="card"
    role="main"
  >
    <div
      class="card-header"
    >
      <h5
        class="card-title"
      >
        Profiles
      </h5>
    </div>
    <div
      class="card-body"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item profileItem"
        >
          <div>
            <label
              for="name-1-input"
            >
              Name
            </label>
            <input
              class="form-control"
              id="name-1-input"
              type="text"
              value="Ben"
            />
          </div>
          <div>
            <label
              for="type-1-select"
            >
              Type
            </label>
            <select
              class="form-control"
              id="type-1-select"
            >
              <option>
                guest
              </option>
              <option>
                user
              </option>
              <option>
                admin
              </option>
            </select>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item profileItem"
        >
          <div>
            <label
              for="name-2-input"
            >
              Name
            </label>
            <input
              class="form-control"
              id="name-2-input"
              type="text"
              value="Sue"
            />
          </div>
          <div>
            <label
              for="type-2-select"
            >
              Type
            </label>
            <select
              class="form-control"
              id="type-2-select"
            >
              <option>
                guest
              </option>
              <option>
                user
              </option>
              <option>
                admin
              </option>
            </select>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
        <li
          class="list-group-item profileItem active"
        >
          <div>
            <label
              for="name-3-input"
            >
              Name
            </label>
            <input
              class="form-control"
              id="name-3-input"
              type="text"
              value=""
            />
          </div>
          <div>
            <label
              for="type-3-select"
            >
              Type
            </label>
            <select
              class="form-control"
              id="type-3-select"
            >
              <option>
                guest
              </option>
              <option>
                user
              </option>
              <option>
                admin
              </option>
            </select>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    >
      <div
        class="wrapper"
      >
        <button
          class="btn btn-primary"
        >
          Create New Profile
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`test delete button 1`] = `
<DocumentFragment>
  <div
    aria-label="Profile Card"
    class="card"
    role="main"
  >
    <div
      class="card-header"
    >
      <h5
        class="card-title"
      >
        Profiles
      </h5>
    </div>
    <div
      class="card-body"
    >
      <ul
        class="list-group"
      >
        <li
          class="list-group-item profileItem"
        >
          <div>
            <label
              for="name-2-input"
            >
              Name
            </label>
            <input
              class="form-control"
              id="name-2-input"
              type="text"
              value="Sue"
            />
          </div>
          <div>
            <label
              for="type-2-select"
            >
              Type
            </label>
            <select
              class="form-control"
              id="type-2-select"
            >
              <option>
                guest
              </option>
              <option>
                user
              </option>
              <option>
                admin
              </option>
            </select>
          </div>
          <div
            class="buttonWrapper"
          >
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div
      class="card-footer"
    >
      <div
        class="wrapper"
      >
        <button
          class="btn btn-primary"
        >
          Create New Profile
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
